### DISCUSSION TITLE
Installation Problems

### DISCUSSION THREAD

**Discussion by LCMichaelKeen** [ucfcdl/UDOIT#707] [Q&A]
Hello,
I'm hoping this is an appropriate place for this post. I'm running into difficulties upgrading to version 3 of UDOIT. I had been running an earlier version on the free Heroku platform. I attempted a new Heroku installation which seems to work on courses with fewer pages, but scans endlessly on larger courses. Even on smaller courses, about 30% of the attempted fixes return errors.

I figured that the free Heroku plan might not have the necessary horsepower to run 3.0. So I'm attempting to follow the documentation to install UDOIT locally on a Linux server.  I'm finding INSTALL.md to be a little vague with regard to what's optional and what's required. For example, it says, "If you are setting up a local UDOIT instance through Docker..." But it doesn't say if there are alternatives to using Docker.  In any event, I think I'm getting hung up at this command:

`docker-compose up --build`

It returns a series of errors saying "Connection aborted" and "Permission Denied."

If I run the command with sudo, I get a bit further but there are lots of warnings, and ultimately, the next command, docker exec -it udoit_php_1 /bin/bash, fails with more permission errors 'while trying to connect to the Docker daemon socket.'

Some guidance here would be tremendously appreciated. If this is an inappropriate post, kindly redirect me.
Thanks.

Michael Keen
Landmark College

**Comment by dtod** [comment]
We use rootless podman (in _theory_ a straight swap for Docker) here, and I've had to spend quite a while wrapping my head around UDOIT, docker, etc.  I think I have it working, but it was quite a struggle - due to my own lack of experience with these things rather than inherent issues in the code. I intend to write it up in full, but here's a high-level overview.

Make sure you have a

`.env.local `

file for starters, and in mine I set `APP_ENV=prod` because weird things happened with dependencies when it was anything else.

I agree about the local Install instructions as far as Docker is concerned. Once I got my docker build working, I did NOT need to do anything with symfony, composer, the database, or yarn.

My specific build command was:

`podman-compose up -d --build --build-arg ENVIORNMENT_TYPE=local 
`
docker-compose should be identical. The `-d` detaches because the php-fpm command doesn't otherwise. The build-arg being local tells various scripts to skip steps e.g. in `deploy/udoit-ng.sh`.

In my specific case with rootless podman, the volume definition in docker-compose.yml for the PHP image was a big stumbling block. It appears to be mounting the udoit folder to /var/www/html, but the Dockerfile does a bunch of COPY statements to that location, which seems contradictory ([to my understanding](https://stackoverflow.com/questions/55197971/docker-copy-command-not-mounting-a-directory)), so I actually commented that out in docker-compose.yml. Also, rootless podman and permissions can be a [real head-scratcher](https://www.tutorialworks.com/podman-rootless-volumes/).

`    #volumes:`
`      #- .:/var/www/html`

I also made a number of changes to get Apache running such as commenting out the following
` RUN apachectl start`
In Dockerfile - that doesn't make sense to me because I don't believe that's how images work. CMD and ENTRYPOINT are for running things that keep running.

I'm not quite sure how to share all of this work to be honest. I'm not very comfortable with how github works. I hope it helps.

I will say that the final image is massive (1.5GB!) mostly, I think, because of the vast numbers of dependencies (node, symfony, yarn, composer, etc.)

**Comment by rob-3** [comment]
Thanks for using UDOIT!

I'm currently in the process of rewriting the Docker configuration to be as close to a one-command install as possible, and also cutting down the image size. ~The free tier of Heroku should have plenty of horsepower for UDOIT 3.~ There may be performance issues with the free tier of heroku on large courses. If you have issues with fixes not working or other errors, please try the `dev/v3-2-0` branch and then [file an issue](https://github.com/ucfopen/UDOIT/issues) or comment here if the problem persists.
